Source/NSubstitute/Core/ICustomHandlers.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.Generic;
+
+namespace NSubstitute.Core
+{
+    /// <summary>
+    ///     Factory method which creates <see cref="ICallHandler" /> from the <see cref="ISubstituteState" />.
+    /// </summary>
+    public delegate ICallHandler CallHandlerFactory(ISubstituteState substituteState);
+
+    public interface ICustomHandlers
+    {
+        IEnumerable<ICallHandler> Handlers { get; }
+        IDictionary<object, object> HandlerDataStorage { get; }
```

Comment:
@dtchepak Many thanks for your example. That is was I was missing. Now it looks obvious, but when I designed API - I didn't see that 😕 
Indeed, we could use closure in factory method to associate context with single substitute:
```csharp
var afState = new AFState();
sub.GetCallRouter().RegisterHandler(subState => new AFHandler(subState, afState));
```

@alexandrnikitin I'd remove this storage a bit later, it isn't required now.